vmtkinteractive.py

```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+## Program:   vmtk
+## Module:    $RCSfile: vmtkinteractive.py,v $
+## Language:  Python
+## Date:      $Date: 2018/04/02 10:45:42 $
+## Version:   $Revision: 1.4 $
+
+##   Copyright (c) Richard Izzo, Luca Antiga. All rights reserved.
+##   See LICENSE file for details.
+
+##      This software is distributed WITHOUT ANY WARRANTY; without even
+##      the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+##      PURPOSE.  See the above copyright notices for more information.
+
+'''
+This module allows for the use of vmtkscripts in an interactive python environment
+(ie. Jupyter Notebook). When importing vmtk, only the 'run' method symbol will be exported.
+Other functions within this module provide for parsing of input arguments, access to the
+local variables of the calling scope, and connection of the vmtk.run method to the
+broader pypes library. 
+'''
+
+import sys
+import os
+import inspect
+from .pypes.pype import PypeRun
+from .pypes.pypesutils import all_indices
+from types import SimpleNamespace
+from functools import wraps
+
+__all__ = ['run']
+
+def locals_to_kwargs(original_function):
+    '''decorator to capture f-string variables of calling scope as kwargs for the PypeRun operation
+    
+    This is designed to enable the execution of traditional pype commands within a jupyter notebook.
+    It identifies the characters enclosed within {curly braces}, and searches the local variables of
+    the calling scope for their values. These values are then inserted into a dictionary with the same
+    key name, and passed as kwargs to the decorated function. 
+
+    This is used in order to simplify the user experience. Instead of having to explicitly pass python 
+    objects into the pype, they can just name them within the pype specification string. 
+
+    This decorator also handles checking for valid class objects and variables identified by the names
+    in {curly braces}. It will raise an error if a python object with the specified name is not alive
+    in the calling scope. 
+    '''
+    @wraps(original_function)
+    def wrapper_func(arguments, **kwargs):
+        '''create the kwarg dictionary of the input argument string mapped to locals of the calling frame'''
+        frame = inspect.currentframe()
+        callingLocals = frame.f_back.f_locals
+        startIndices = all_indices('{', arguments)
+        endIndices = all_indices('}', arguments)
+        zippedIndices = zip(startIndices, endIndices)
```

Comment:
check what to do in case curly braces are unbalanced (odd or in a weird sequence):
you have to check that every start index precedes the corresponding end index and that each end index precedes the next start index.
you could interleave start and end indices, make a sorted copy of the resulting list and compare the two. If it's a match, parentheses are balanced (if start and end indices have equal length)